refactor: migrate account module to TS

Kalovelo opened the pull request:
This PR creates declaration files for stampit plugins along with types for the account module.
Since Stampit types do not exist on definitely typed, a manual approach was used by generating declaration files from the official repository.

- A new types package for chai was added in order to also migrate tests in TS
- A tsconfig change occured which is based on stampit's importing issue. `esModuleInterop` was set to `true`. Read more: https://www.typescriptlang.org/tsconfig#esModuleInterop
- Types are stored in `[MODULE].types.ts` in order to pair them with the corresponding module file `[MODULE].ts`
- Stamp function JSDocs were moved in the corresponding types files in order to store docs&typings in one file and business logic to another
- some `// @ts-ignore` comments were added where necessary until types for those imported files are created.

subhod-i commented on src/account/base.ts:
```diff
@@ -0,0 +1,130 @@
+  */
+
+export default stampit<AccountBase>({
+  init (this: AccountBase, { networkId }) { // NETWORK_ID using for signing transaction's
```
No type for the options.

davidyuk reviewed (commented):
I think would be more efficient to drop stamps firstly, otherwise TypeScript looks quite tricky and we will have to rewrite it later while dropping stamps.

Also, I'm asking to:
- use "refactor" prefix instead of "feat" in commit messages to don't get these commits in release notes
- avoid internal merge commits (rebase at the top of develop instead) to have history clearer
- avoid overlapping changes between commits (like committing unformatted code and fixing it later)

Kalovelo commented:
Thank you for the detailed review @davidyuk 

1. interfaces have been moved to the corresponding files
2. `_resolveAccount` throwing error has been set to the original one
3. Fixed eslint not throwing error by setting maxLen string ignoring to false. This is useful in anycase since long strings are  not readable either
4. Created `TxOptions` interface based on docs
5. Replaced `Parameter<...>`
6. Set `AccountMultiple.accounts` as an object since is has indexed fields. This allowed us to remove the explicit casting

subhod-i commented on src/account/base.ts:
```diff
@@ -0,0 +1,131 @@
+  * @return {Boolean}
+  */
+async function verifyMessage (this:AccountBase, message: Parameters<AccountBase['verifyMessage']>[0], signature: Parameters<AccountBase['verifyMessage']>[1], opt: Parameters<AccountBase['verifyMessage']>[2] = {}) {
+  return verifyMessageCrypto(
```
@Kalovelo can use the `npm run lint` script locally to avoid such errors being thrown from the GitHub workflow.

subhod-i commented:
@Kalovelo linting errors https://github.com/aeternity/aepp-sdk-js/runs/6060166544?check_suite_focus=true#step:7:10

Kalovelo commented:
I think it looks good

- We could consider removing the underscore from classes since they're both exported and extended therefore not private and this might confuse newer contributors.
- Init function as a constructor seems like a great bridge between classes & stamps, however, I think we could document it somewhere that they will be transformed in constructors later on and it is a workaround
- Like you commented before, we could reduce the use of the `Parameters<...>` convention by creating initialization options interfaces for each account Instance since this could make it more readable
- Finally, as you proposed in a previous comment, we could create A `TxOptions` interface in order to reduce the both the use of `Parameters` and also have a single source of truth for the Transaction Options

davidyuk commented:
> We could consider removing the underscore from classes since they're both exported and extended therefore not private and this might confuse newer contributors

I use underscore to show that these classes are not a part of public api, e.g. not intended for external using (until we drop stamps), and we won't consider changing them as a breaking change. I think we can keep the naming as it is untill we remove account stamps. Alternatively, we can name them like `AccountBaseInternal`.

> document it somewhere that they will be transformed in constructors later on

I've added a comment in `_AccountBase`

> we could reduce the use of the `Parameters<...> `

I've noticed that TypeScript forces the inheritor methods to accept the same arguments as the base class, so we can explicitly define arguments types each time and don't worry to make a typo. From the other side, parent class can have arguments that are not related to the inheritor's method, in this case I would rather point to the type of base class. We can do it using `Parameters<...>` or by creating an alias type. The first approach looks like a more explicit inheritance, the second potentially makes us to edit the base class while overriding it's methods. So I would like to keep it as it is.

> we could create A `TxOptions` interface

I saw it in your version, but I forgot to explain why I dropped it. The thing is that it were explaining options that account methods doesn't accepts, like `fee`, `nonce` (txBuilder); `onAccount` (account multiple/account resolver); `innerTx` (`signTransaction`); `verify`, `waitMined` (`sendTransaction`). I think instead of defining an object that would specify all possible options would be better to combine options of the current method with options of methods that are called, like:
```ts
async function spend (
  amount: number | string,
  recipientIdOrName: string,
  options: Parameters<this.send>[1] & Parameters<this.spendTx>[0] &
    Parameters<this.address>[0] & Parameters<this.resolveName>[2]
) {
  const opt = { ...this.Ae.defaults, ...options }
  return this.send(
    await this.spendTx({
      ...opt,
      senderId: await this.address(opt),
      recipientId: await this.resolveName(recipientIdOrName, 'account_pubkey', opt),
      amount
    }),
    opt
  )
}
```
This way user would get more accurate list of options. I'm not sure that it can be made more readable using type aliases or so.

davidyuk commented on src/utils/encoder.ts:
```diff
@@ -68,11 +68,11 @@ const base58 = {
   if (requiredPrefix != null && requiredPrefix !== prefix) {
     throw new PrefixMismatchError(prefix, requiredPrefix)
   }
+  if (encodedPayload == null) throw new DecodeError(`Encoded string missing payload: ${data}`)
```
Yep. While TS checks that string starts with a specific prefix, I would keep the runtime check for non-ts environments, at least for the nearest future

davidyuk commented on test/unit/memory-account.js:
```diff
@@ -20,29 +20,20 @@ import { describe, it } from 'mocha'
 import { expect } from 'chai'
 import MemoryAccount from '../../src/account/memory'
 import { generateKeyPair } from '../../src/utils/crypto'
-import { InvalidKeypairError } from '../../src/utils/errors'
```
Test files also getting migrated to ts, this one will be migrated in next PRs
